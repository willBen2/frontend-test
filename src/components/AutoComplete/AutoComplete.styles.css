.auto-complete-container {
    position: relative;
    display: inline-block;
}

.dropdown-container {
    border: 1px solid #c9c9c9;
    border-bottom: none;
    display: block;
    margin: 0;
    z-index: 1050;
    box-sizing: border-box;
    overflow: hidden;
    font-size: 14px;
    background-color: #fff;
    border-radius: 2px;
    outline: none;
    width: 100%;
    flex-direction: column;
}

.dropdown-item {
    padding: 10px;
    border-bottom: 1px solid #c9c9c9;
}

.dropdown-item:hover {
    background-color: #dedede;
    cursor: pointer;
}

.highlight {
    color: cornflowerblue;
    font-weight: 600;
}
