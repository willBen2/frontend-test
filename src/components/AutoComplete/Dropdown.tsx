import React, { ReactNode } from 'react';

interface DropdownProps {
  children: ReactNode[];
  className?: string;
}

export default function Dropdown({ children, className, ...rest }: DropdownProps) {
  if (!children || !children.length) {
    return null;
  }
  return (
    <div className={["dropdown-container", className].join(' ')} {...rest}>
      {children}
    </div>
  )
}
