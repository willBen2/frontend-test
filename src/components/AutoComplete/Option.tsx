import React from 'react';
import Text from '~/components/Text';

export type AutoCompleteOptionType = {
  title: string;
  search?: string;
  className?: string;
  onSelect: (title: string) => void;
}

export default function Option(props: AutoCompleteOptionType) {
  const { title, onSelect = () => {}, search } = props;
  const onClick = () => {
    onSelect(title);
  }

  let lTitle = title.toLowerCase();
  let lSearch = search?.toLowerCase()!;

  let start = title;
  let match = '';
  let end = '';
  if (search) {
    const index = lTitle.indexOf(lSearch);
    start = title.slice(0, index);
    match = title.slice(index, index + search.length);
    end = title.slice(index + search.length);
  }

  return (
    <div
      onClick={onClick}
      className={[
        "dropdown-item",
        lSearch===lTitle ? 'dropdown-item-match' : '',
        props.className
      ].join(' ')}
    >
      <Text className="option-title">
        {start}<Text className="highlight">{match}</Text>{end}
      </Text>
    </div>
  )
}
