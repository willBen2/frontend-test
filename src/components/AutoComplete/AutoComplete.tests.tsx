import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import AutoComplete from './AutoComplete';

const mockOptions = [
  {
    title: 'test',
    id: 1,
  },
  {
    title: 'testing',
    id: 2,
  },
  {
    title: 'tes t',
    id: 3,
  },
  {
    title: 't es t',
    id: 4,
  },
]

describe('AutoComplete', () => {
  it('should match with default AutoComplete', () => {
    const onChangeMock = jest.fn();
    const { container } = render(
      <AutoComplete  onSearchChange={onChangeMock} options={mockOptions} />
    )
    expect(container).toMatchSnapshot();
  });

  it('should not show dropdown', () => {
    const onChangeMock = jest.fn();
    render(
      <AutoComplete
        dropdownTestId="dropdown"
        options={mockOptions}
        onSearchChange={onChangeMock}
      />
    );

    const dropdowns = screen.queryAllByTestId('dropdown');
    expect(dropdowns.length).toBe(0);
  });

  it('should show dropdown', () => {
    const onChangeMock = jest.fn();
    render(
      <AutoComplete
        inputTestId="input"
        dropdownTestId="dropdown"
        options={mockOptions}
        onSearchChange={onChangeMock}
      />
    );

    fireEvent.focus(screen.getByTestId('input'));

    const dropdown = screen.getByTestId('dropdown');
    expect(dropdown).toBeDefined();
  });

  it('should fire onSearchChange', () => {
    const onChangeMock = jest.fn();
    render(
      <AutoComplete
        inputTestId="input"
        options={mockOptions}
        onSearchChange={onChangeMock}
      />
    );

    fireEvent.change(screen.getByTestId('input'), {
      target: { value: 'test' },
    });

    expect(onChangeMock).toHaveBeenCalled();
  });

  it('should have two items', () => {
    const onChangeMock = jest.fn();
    const { container } = render(
      <AutoComplete
        inputTestId="input"
        dropdownTestId="dropdown"
        options={mockOptions}
        onSearchChange={onChangeMock}
      />
    );

    fireEvent.focus(screen.getByTestId('input'));
    fireEvent.change(screen.getByTestId('input'), {
      target: { value: 'test' },
    });

    const dropdown = screen.getByTestId('dropdown');
    expect(dropdown.children.length).toBe(2);
  });
})
