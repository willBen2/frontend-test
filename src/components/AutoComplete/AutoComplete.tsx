import React, { HTMLAttributes, useEffect, useState } from 'react';
import { SearchBar } from '~/components';
import './AutoComplete.styles.css';
import Dropdown from '~/components/AutoComplete/Dropdown';
import Option from '~/components/AutoComplete/Option';

interface AutoCompleteProps {
  options: {title: string, id: number | string}[];
  inputTestId?: string;
  dropdownTestId?: string;
  placeholder?: string;
  inputClassname?: string;
  dropdownClassName?: string;
  dropdownItemClassName?: string;
  onSearchChange: (search: string) => void;
}

function AutoComplete(props: AutoCompleteProps & HTMLAttributes<HTMLElement>) {
  const [search, setSearch] = useState<string>('');
  const [showDropdown, setShowDropdown] = useState(false);
  const { onSearchChange, options } = props;

  const onSelect = (s: string) => {
    setShowDropdown(false);
    setSearch(s);
  };

  const onChange = (s: string) => {
    setSearch(s);
  };

  const onFocus = () => {
    setShowDropdown(true);
  };

  useEffect(() => {
    onSearchChange(search);
  }, [search]);

  return (
    <div className={["auto-complete-container", props.className].join(' ')}>
      <SearchBar onSearchChange={onChange} search={search}>
        <SearchBar.Input onFocus={onFocus} data-testid={props.inputTestId} placeholder={props.placeholder} className={props.inputClassname} />
      </SearchBar>
      {
        showDropdown
          ? <Dropdown data-testid={props.dropdownTestId} className={props.dropdownClassName}>
            {
              options.filter(o => o.title.toLowerCase().includes(search.toLowerCase())).map(option => (
                <Option
                  key={option.id}
                  className={props.dropdownItemClassName}
                  onSelect={onSelect}
                  search={search} title={option.title}
                />
              ))
            }
          </Dropdown>
          : null
      }
    </div>
  );
}

export default AutoComplete;
