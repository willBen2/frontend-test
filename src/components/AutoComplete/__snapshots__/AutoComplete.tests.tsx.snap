// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoComplete should match with default AutoComplete 1`] = `
<div>
  <div
    class="auto-complete-container "
  >
    <div
      class="input-container"
    >
      <input
        class="input"
        value=""
      />
    </div>
  </div>
</div>
`;
